Add methods to Plugin classes fix issue #26

aldebaran30701 opened the pull request:
- PluginParameter setPropertyWithoutEvent
 - SingleChoiceParameterType setOptionsWithoutEvent
 - usage added within plugins
   - ExtractWordsFromText
   - SplitNodesPlugin
Implementation should fix issue #26

<!--

### Requirements

* Filling out the template is required. Any pull request that does not include 
enough information to be reviewed in a timely manner may be closed at the 
maintainers' discretion.
* All new code requires unit tests to ensure they work as expected and will 
continue to work as new code is added in the future (regression testing).
* Have you read Constellation's Code of Conduct? By filing an issue, you are 
expected to comply with it, including treating everyone with respect: 
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Description of the Change
This change adds methods to assist in the management and configuration of plugins in the DAV. The methods allow the plugins to set parameters and options which may depend on the type of graph opened and within focus. this enables the plugin to configure it's options without enabling itself through a changeEvent triggering the plugin to activate.
The method name has _WithoutEvent concatenated onto it which allows for ease of readability and less confusion of an overloaded method. the method uses the same logic from the original method to ensure consistency in it's operation.
The methods in this PR have been implemented within the following plugins which had issues relating to #26. 
<!--

We must be able to understand the design of your change from this description. 
If we can't get a good idea of what the code will be doing from the description 
here, the pull request may be closed at the maintainers' discretion. Keep in 
mind that the maintainer reviewing this PR may not be familiar with or have 
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs
Overloaded method signatures with a boolean flag left the same amount of code in each file, but reduced readability. The original idea was to tackle the issue within setChoice which was commented out previously. This was a two part issue, which both appear to be solved with this fix.
<!-- 

Explain what other alternates were considered and why the proposed version was 
selected.

-->

### Why Should This Be In Core?
This should be in core as it defines the "expected" behaviour of loading a graph when the DAV is opened. It solves issue #26 and allows for future expansion with any other plugin which may need to configure settings dependent on the graph loaded.
<!--

Explain why this functionality should be in Constellation Core as opposed to a 
different module suite.

-->

### Benefits
This fix as mentioned before, allows for future plugin development and also enables the user to create a graph without having to deselect the automatically enabled plugins.
<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks
None realized thus far.
<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process
Run all tests in suite
Open Constellation
 - Open DAV -> open graph = no plugin automatically activated.
- Open DAV -> open graph -> open SplitNodesPlugin -> select an option from dropdown -> deselect enabled checkbox -> close graph -> open new graph = no plugin automatically activated.
- Open DAV -> open graph -> draw two nodes -> change identifiers -> split by identifier -> close graph -> deselect plugin -> new graph = no plugin automatically activated.
- Open DAV ->open graph -> open both SplitNodesPlugin and ExtractWordsFromTextPlugin and select parameters -> deselect both plugins ->close graph -> new graph = no plugin automatically activated.
- Open DAV ->open graph -> close graph -> close DAV -> open DAV -> open graph = no plugin automatically activated.
<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., buttons you clicked, text you typed, 
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
N/A
<!-- Enter any applicable Issues here -->

arcturus2 commented:
@aldebaran30701 this works and I'm happy with it.

@cygnus-x-1 if you have time, I'd like your feedback on this.

cygnus-x-1 commented:
This might work for the two plugins identified, however there are several other methods in SingleChoiceParameterType and MultiChoiceParameterType (and maybe other ParameterType classes) which fire change events and do have not been provided an equivalent fix. I think a proper solution to this should be built into the framework, not just this SingleChoiceParameterType.

arcturus2 commented:
Thanks @cygnus-x-1, your feedback is always valued.

@aldebaran30701 would you be able to think about this a bit more?

I've been trying to think about a viable solution and I'm wondering if something like this is a good idea? It would be to suppress the event firing by introducing a `suppressEvent()` method to `PluginParameter`? The `fireChangeEvent()` is part of `PluginParameter` so it will be easy to not fire.

Then, you can use it like this in the case of `updateParameters()` and add some docs to `Plugin.updateParameters()` to suggest it would be useful to use `PluginParameter.suppressEvent()` and explain the reason.

```java
transactionType.suppressEvent(true);
if (types.contains("Correlation")) {
    SingleChoiceParameterType.setChoice(transactionType, "Correlation");
}
transactionType.suppressEvent(false);
```

aldebaran30701 commented:
I really like that idea @arcturus2 
Allows for consistent practice throughout the whole code base if needed for other items firing events.
I can start working on something today, following along the lines of what you have outlined above.
@cygnus-x-1 do you for see any issues with this implementation?

arcturus2 commented:
Good thinking @aldebaran30701. Can we have both method signatures? 😃 
If you don't pass anything then it suppresses everything, otherwise you can specify your list.

aldebaran30701 commented:
done and done. 
I've updated the changelog, added documentation to plugin.updateparameters().
used a collection to hold the currently "filtered" pluginparameters. 
used a boolean flag to check whether or not to filter a result. 
I have one other improvement I would like to make, that would be to refactor the checking of the events, into the fire change event itself.

aldebaran30701 commented:
Fixed. Much neater.

arcturus2 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/pluginframework/parameters/PluginParameter.java:
```diff
@@ -70,6 +71,11 @@
+    
+    private ArrayList<ParameterChange> suppressedEvents = new ArrayList<>();
+    
+    private static final ArrayList<ParameterChange> PARAMETER_EVENTS = new ArrayList<>(Arrays.asList(ParameterChange.values()));
```
Ditto, can you please change to `private static final List`. It doesn't make too much difference, its more my OCD 😸 .

arcturus2 replied:
Do you think renaming it to `ALL_PARAMETER_EVENTS` makes for better understanding?

arcturus2 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/pluginframework/parameters/types/SingleChoiceParameterType.java:
```diff
@@ -170,6 +170,7 @@ public static ParameterValue getChoiceData(final PluginParameter<SingleChoicePar
 
     /**
      * Set the current selected choice from a string.
+     * This will fire a {@link ParameterChange} event if successful.
```
If you are going to add this to `SingleChoiceParameterType` can you add it to the other *ParameterType classes for consistency? Otherwise removing it is also OK.

arcturus2 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/pluginframework/parameters/PluginParameter.java:
```diff
@@ -150,6 +156,38 @@ public void setVisible(boolean visible) {
+     * when specifying all events or enabling events. Pass {@link ParameterChange} 
+     * enumerated types when specifying certain events to suppress.
+     */
+    public void suppressEvent(final boolean suppress, final ArrayList<ParameterChange> eventsToSuppress){
```
Please change the parameter type to `List`.

I might be alone in thinking that it feels a little weird to specify an empty list to suppress everything but passing a populated list to ignore specific ones. Perhaps we could have another method being `public void suppressEvent(final boolean suppress)` which will just call this `suppressEvent` with the empty list.

Happy to be told otherwise.

aldebaran30701 replied:
passing a populated list should suppress only the ParameterChange events listed within it.
I've done the other changes and pushed them, but yet to change the suppressEvent(...)

arcturus2 replied:
> passing a populated list should suppress only the ParameterChange events listed within it.

Yep, so if we go with that logic then passing an empty list should suppress nothing right? But it suppresses everything. I'm OK with this though because there is javadoc explaining it. My only minor suggestion is that we could have another method signature of `public void suppressEvent(final boolean suppress)` which will hide the detail of needing to pass the empty list.
